hbase-handler/src/java/org/apache/hadoop/hive/hbase/HiveHBaseTableInputFormatV2.java

```diff
@@ -0,0 +1,370 @@
+import org.apache.hadoop.hbase.mapred.TableMapReduceUtil;
+import org.apache.hadoop.hbase.mapreduce.TableSplit;
+import org.apache.hadoop.hive.hbase.ColumnMappings.ColumnMapping;
+import org.apache.hadoop.hive.ql.exec.SerializationUtilities;
+import org.apache.hadoop.hive.ql.index.IndexPredicateAnalyzer;
+import org.apache.hadoop.hive.ql.index.IndexSearchCondition;
+import org.apache.hadoop.hive.ql.plan.ExprNodeDesc;
+import org.apache.hadoop.hive.ql.plan.ExprNodeGenericFuncDesc;
+import org.apache.hadoop.hive.ql.plan.TableScanDesc;
+import org.apache.hadoop.hive.serde.serdeConstants;
+import org.apache.hadoop.hive.serde2.SerDeException;
+import org.apache.hadoop.hive.serde2.typeinfo.TypeInfo;
+import org.apache.hadoop.hive.serde2.typeinfo.TypeInfoUtils;
+import org.apache.hadoop.hive.shims.ShimLoader;
+import org.apache.hadoop.mapred.InputFormat;
+import org.apache.hadoop.mapred.InputSplit;
+import org.apache.hadoop.mapred.RecordReader;
+import org.apache.hadoop.mapred.JobConf;
+import org.apache.hadoop.mapred.Reporter;
+import org.apache.hadoop.mapreduce.Job;
+import org.apache.hadoop.mapreduce.JobContext;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+import org.apache.hadoop.mapreduce.lib.input.FileInputFormat;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.security.PrivilegedExceptionAction;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * It`s the pair with {@link HiveHBaseTableInputFormat}, instead of directly extending TableInputFormatBase, using delegate
+ * to make the class ONLY inherit from mapred.*, which makes the hierarchy more clear and avoid downstream application
+ * like spark issue, ex https://github.com/apache/spark/pull/31302
+ */
+public class HiveHBaseTableInputFormatV2 implements InputFormat<ImmutableBytesWritable, ResultWritable> {
+
+  static final Logger LOG = LoggerFactory.getLogger(HiveHBaseTableInputFormatV2.class);
+  private static final Object HBASE_TABLE_MONITOR = new Object();
+
+  private HiveHBaseTableInputFormatDelegate delegate = new HiveHBaseTableInputFormatDelegate();
+
+  @Override public RecordReader<ImmutableBytesWritable, ResultWritable> getRecordReader(InputSplit split,
+                                                                                        JobConf jobConf, final Reporter reporter) throws IOException {
+
+    HBaseSplit hbaseSplit = (HBaseSplit) split;
+    TableSplit tableSplit = hbaseSplit.getTableSplit();
+
+    final org.apache.hadoop.mapreduce.RecordReader<ImmutableBytesWritable, Result> recordReader;
+
+    Job job = new Job(jobConf);
+    TaskAttemptContext tac = ShimLoader.getHadoopShims().newTaskAttemptContext(job.getConfiguration(), reporter);
+
+    final Connection conn;
+
+    synchronized (HBASE_TABLE_MONITOR) {
+      conn = ConnectionFactory.createConnection(HBaseConfiguration.create(jobConf));
```

Comment:
This connection is not in try block and I do not see where this connection is closed. 
Only the catch contains close. Maybe just because of the synchronized, but hard to follow.

Reply:
The connection is used by the RecordReader, so the close will be involved in the reader close() which at line 107